perfkitbenchmarker/benchmarks/cloudsuite_web_serving.py

```diff
@@ -0,0 +1,411 @@
+  client.RemoteCommand('perl -pi -e '
+                       '"s/FRONTEND_IP/%s/g"'
+                       ' %s/run.xml' % (FRONTEND_IP, FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/MYSQL_DIR/%s/g" '
+                       '%s/run.xml' % (re.escape(MYSQL_HOME), FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/OUTPUT_DIR/%s/g" '
+                       '%s/run.xml' % (re.escape(OUTPUT_DIR), FABAN_HOME))
+  filestore = posixpath.join(frontend.GetScratchDir(), 'filestore')
+  client.RemoteCommand('perl -pi -e '
+                       '"s/FILESTORE_DIR/%s/g" '
+                       '%s/run.xml' % (re.escape(filestore), FABAN_HOME))
+  client.RemoteCommand('perl -pi -e '
+                       '"s/BACKEND_IP/%s/g" '
+                       '%s/run.xml' % (BACKEND_IP, FABAN_HOME))
+  return
+
+
+def GetInfo():
+  return BENCHMARK_INFO
+
+
+def CheckPrerequisites():
+  """Verifies that the required resources are present.
+  Raises:
+    perfkitbenchmarker.data.ResourceNotFound: On missing resource.
+  """
+  if FLAGS['num_vms'].present and FLAGS.num_vms < 3:
+    raise ValueError('Web Serving requires at least 3 VMs')
+  return
+
+
+def PreparePrivateKey(vm):
+  vm.AuthenticateVm()
+
+
+def Prepare(benchmark_spec):
+  """Install Java, apache ant
+     Set up the client machine, backend machine, and frontend
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  vms = benchmark_spec.vms
+  vm_util.RunThreaded(PreparePrivateKey, vms)
+  client = vms[CL]
+  backend = vms[BK]
+  frontend = vms[FR]
+  CLIENT_IP = client.ip_address
+  for vm in vms:
+    vm.Install('wget')
+    vm.Install('ant')
+    vm.Install('openjdk7')
+  client.RemoteCommand('cd %s && '
+                       'wget parsa.epfl.ch/cloudsuite/software/web.tar.gz && '
+                       'tar xzf web.tar.gz' % vm_util.VM_TMP_DIR)
+  backend.RemoteCommand('cd %s && '
+                        'wget parsa.epfl.ch/cloudsuite/software/web.tar.gz && '
+                        'tar xzf web.tar.gz' % vm_util.VM_TMP_DIR)
+  frontend.RemoteCommand('cd %s && '
+                         'wget parsa.epfl.ch/cloudsuite/software/web.tar.gz && '
+                         'tar xzf web.tar.gz' % vm_util.VM_TMP_DIR)
+  frontend.RemoteCommand('cd %s && '
+                         'wget parsa.epfl.ch/cloudsuite/software/perfkit/'
+                         'web_serving/webservingfiles.tgz && '
+                         'tar xzf webservingfiles.tgz' % BASE_DIR)
+  setupClient(benchmark_spec)
+  time.sleep(30)
+  frontend.RemoteCommand('scp -r -o StrictHostKeyChecking=no %s:%s %s' %
+                         (CLIENT_IP, FABAN_HOME, BASE_DIR))
+  backend.RemoteCommand('scp -r -o StrictHostKeyChecking=no %s:%s %s' %
+                        (CLIENT_IP, FABAN_HOME, BASE_DIR))
+  setupBackend(benchmark_spec)
+  setupFrontend(benchmark_spec)
```

Comment:
Calling these setup functions may also be parallelizable, although I admit our existing RunThreaded function is not designed ideally for executing different targets. That said, considering how complicated the setups are, I think this could take a significant cut out of the benchmark prepare stage.